src/main/java/gov/usgs/earthquake/distribution/WebSocketClient.java

```diff
@@ -10,121 +10,105 @@
     this.endpoint = endpoint;
-    this.timeoutMillis = timeoutMillis;
+    this.reconnect = reconnect;
+    this.reconnectInterval = reconnectInterval;
   }
 
 
   public void startup() {
     stopThread = false;
-    this.thread = new Thread(this);
-    this.thread.start();
+    thread = new Thread(this);
+    thread.start();
   }
 
   public void shutdown() throws Exception {
     // bring down thread
     stopThread = true;
-    thread.interrupt();
+    synchronized (close) {
+      thread.interrupt();
+    }
     thread.join();
+    thread = null;
+  }
+
+  public void connect() throws Exception {
+    // try to connect to server
+    WebSocketContainer container = ContainerProvider.getWebSocketContainer();
+    container.connectToServer(this, endpoint);
+  }
+
+  public void disconnect() throws Exception {
+    if (isConnected()) {
+      session.close();
+    }
   }
 
   public void run() {
-    synchronized (sync) {
-      while (!stopThread) {
-        doReconnect = false;
-        try {
+    while (!stopThread) {
+      // try to connect
+      try {
+        synchronized(close) {
           connect();
-          if (!doReconnect) {
-            sync.wait();
-          }
-        } catch (InterruptedException e1) {
-          // likely told to stop
-          stopThread = true;
-        } catch (Exception e2) {
-          // unresolvable problem connecting; try to resolve with listener
           try {
-            listener.onConnectException(e2);
-          } catch (Exception e3) {
-            // unable to handle; quit
+            // successful connection
+            close.wait();
+          } catch (InterruptedException ie) {
+            // gracefully close, disconnect if possible
             stopThread = true;
+            disconnect();
           }
         }
-      }
-
-      // make sure we're disconnected
-      if (this.session != null && this.session.isOpen()) {
+      } catch (Exception e1) { //exception on connect
+        // let listener handle if they want
         try {
-          this.session.close();
-        } catch (IOException e) {
-          // ignore IOException; just close
+          listener.onConnectException(e1);
+        } catch (Exception e2) {
+          // kill thread on listener exception
+          break;
         }
-      }
-    }
-    listener.onDisconnect();
-  }
 
-  public void connect() throws Exception {
-    // try to connect to server
-    WebSocketContainer container = ContainerProvider.getWebSocketContainer();
-    try {
-      container.connectToServer(this, endpoint);
-    } catch (Exception e1) {
-      try {
-        this.onConnectException(e1);
-      } catch (Exception e2) {
-        throw e2;
+        // reconnect for unhandled connect exception
+        if (!stopThread && reconnect) {
+          try {
+            Thread.sleep(reconnectInterval);
+          } catch (InterruptedException ie) {
+            // gracefully close
+            stopThread = true;
```

Comment:
Do not modify this flag in the thread, it's either set by shutdown or the thread was interrupted for a different reason.

Reply:
removed all stopThread modifications in run()